contracts/Strategy.sol

```diff
@@ -53,80 +222,624 @@ contract Strategy is BaseStrategy {
-        if (_amountNeeded > totalAssets) {
-            _liquidatedAmount = totalAssets;
-            _loss = _amountNeeded.sub(totalAssets);
+                for (
+                    uint256 i = 0;
+                    i < kashiPairs.length &&
+                        sharesToFreeFromKashi > sharesFreedFromKashi;
+                    i++
+                ) {
+                    if (i == lowestInterestIndex) continue; // we already visited this
+
+                    sharesFreedFromKashi = sharesFreedFromKashi.add(
+                        liquidateKashiPair(
+                            i,
+                            sharesToFreeFromKashi.sub(sharesFreedFromKashi)
+                        )
+                    );
+                }
+
+                bentoBox.withdraw(
+                    BIERC20(address(want)),
+                    address(this),
+                    address(this),
+                    0,
+                    sharesInBento()
+                );
+            }
+
+            _liquidatedAmount = balanceOfWant();
+
+            if (_amountNeeded > _liquidatedAmount) {
+                _loss = _amountNeeded.sub(_liquidatedAmount);
+            }
         } else {
             _liquidatedAmount = _amountNeeded;
         }
     }
 
-    function liquidateAllPositions() internal override returns (uint256) {
-        // TODO: Liquidate all positions and return the amount freed.
+    function liquidateAllPositions()
+        internal
+        override
+        returns (uint256 _liquidatedAmount)
+    {
+        (_liquidatedAmount, ) = liquidatePosition(type(uint256).max);
+    }
+
+    // new strategy **must** have the same kashiPairs attached
+    function prepareMigration(address _newStrategy) internal override {
+        for (uint256 i = 0; i < kashiPairs.length; i++) {
+            if (kashiPairs[i].pid != 0) {
+                masterChef.withdraw(
+                    kashiPairs[i].pid,
+                    kashiFactionInMasterChef(i)
+                );
+            }
+
+            kashiPairs[i].kashiPair.transfer(
+                _newStrategy,
+                kashiFractionInPair(i)
+            );
+        }
+    }
+
+    function addKashiPair(address _newKashiPair, uint256 _newPid)
+        external
+        onlyGovernance
+    {
+        // cannot exceed max pair length
+        require(kashiPairs.length < MAX_PAIRS);
+        // must use the correct bentobox
+        require(
+            address(IKashiPair(_newKashiPair).bentoBox()) == address(bentoBox)
+        );
+        // kashPair asset must match want
+        require(IKashiPair(_newKashiPair).asset() == BIERC20(address(want)));
+        if (_newPid != 0) {
+            // masterChef pid token must match the kashiPair
+            require(
+                address(masterChef.poolInfo(_newPid).lpToken) == _newKashiPair
+            );
+        }
+
+        for (uint256 i = 0; i < kashiPairs.length; i++) {
```

Comment:
just a suggestion since this for length is not that big [5], but a mapping address => uint256 == 1 exists or bool may be helpful to check against the set of known pairs.

But if you are running into size issue it may be overkill

Reply:
we are getting close and i'm not sure it's worth it. Wdyt?

```============ Deployment Bytecode Sizes ============
  Strategy       -  23,723B  (96.53%)
  BoringMath     -      86B  (0.35%)
  BoringMath128  -      86B  (0.35%)
  BoringMath32   -      86B  (0.35%)
  BoringMath64   -      86B  (0.35%)
  Address        -      86B  (0.35%)
  Math           -      86B  (0.35%)
  SafeERC20      -      86B  (0.35%)
  SafeMath       -      86B  (0.35%)
  RebaseLibrary  -      86B  (0.35%)```

Reply:
yeah, i think this is fine